src/main/java/com/google/devtools/build/lib/syntax/MethodLibrary.java

```diff
@@ -102,6 +105,51 @@ private static Object findExtreme(
     }
   }
 
+  private static Object evalKey(
+      Object obj, Object key, Location loc, Environment env, FuncallExpression ast)
+      throws EvalException, InterruptedException {
+    if (key == Runtime.NONE) {
+      return obj;
+    } else if (key instanceof BuiltinCallable) {
+      return ((BuiltinCallable) key).call(Arrays.asList(obj), Collections.emptyMap(), ast, env);
+    } else if (key instanceof BaseFunction) {
+      return ((BaseFunction) key).call(Arrays.asList(obj), null, null, env);
```

Comment:
The two cases for BuiltinCallable and BaseFunction look redundant. Can we combine them into a single instanceof test for SkylarkCallable?

Reply:
Neither of them is an instance of SkylarkCallable, and I can't think of an easy way to combine them as the call method signature is different.

Reply:
Sorry, I meant SkylarkFunction, not SkylarkCallable. (A SkylarkFunction represents a value that is callable from Skylark, even though it might not be a function defined in Skylark. The names are unfortunate.)

Reply:
Done, thanks.